rfcs/0000-expressive-diagnostics.md

```diff
@@ -0,0 +1,356 @@
+    the function signature? For example: \n`fn \
+    {fn_name}<{comma_separated_generic_params}>({args}) -> ... `
+```
+
+The way how we structure those messages is, however, not standardized. E.g., here is an another attempt to provide a detailed error explanation, formated and conveyed in a completely different way:
+
+```
+expected: {expected} \n\
+    found:    {given} \n\
+    help:     The definition of this {decl_type} must \
+    match the one in the {interface_name} declaration.
+```
+
+At the moment, although detailed and helpful, these messages are "packed" within a single piece of text pointing to a single place in code. This significantly limits amount of helpful information that can be provided to the programmer. It also limits the presentation of the context in which the diagnostic occurs. The overall context usually spans accross several points in code. We want to be able to place the useful information on the exact places in code that are relevant to the diagnostic.
+
+Sway programmers are also confronted with short, sometimes cryptic error messages, containing jargon used by compiler developers. E.g.:
+
+```
+Invalid value \"{value}\"
+```
+
+```
+Constant requires expression.
+```
+
+Internally, expressive diagnostics will empower Sway language developers to easily define detailed and helpful diagnostics that will be focused on the exact code that the programmer wrote. Diagnostics API will also provide a unified way to define diagnostics. (At the moment we have three different approaches for defining diagnostics. `CompileError`, `CompileWarning`, and `ParseError` are representatives of those different approaches.)
+
+Externally, we expect to see the effects experienced in other communities:
+
+- better perception of Sway as a language
+- easier language learning curve
+- increased efficiency in resolving compiler errors and warnings.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+When using `forc` via command line Sway programmers will get detailed diagnostics consisting of the following elements:
+
+- _Level_: Error or Warning.
+- _Code_: Unique code of the diagnostic.
+- _Reason_: Short description of the diagnostic, not related to a specific issue that the programmer did. The _reason_ answers the question "Why is this an error or warning?" E.g., Because - "Constants cannot be shadowed". It points out the general language rule that was violated.
+- _Issue_: Short description of the concrete issue caused by the programmer. E.g., "Variable "X" shadows imported constant with the same name"
+- _Hints_: Detailed descriptions of the diagnostic. Hints are placed in the source code. They point to the problematic line of code but also to other places in code that can help to understand the issue.
+- _Help_: Additional friendly information, which is not related to a place in code, that helps understanding and solving the issue.
```

Comment:
Which of these are optional, which of these are mandatory?

Reply:
At the moment, for backward compatibility, only the _Level_ and the _Issue_. And since switching to `Diagnostic` will take time, it will stay like that for a while.

Once all errors and warnings support `Diagnostic` IMO these should be mandatory:
- Level
- Code
- Reason
- Issue